File: backend/app/api/laboratorio.py
"""
Router de laboratorio dental.
- CRUD de laboratorios (catálogo)
- CRUD de trabajos de laboratorio
"""
import uuid
from datetime import date
from typing import Annotated

from fastapi import APIRouter, Depends, HTTPException, Query, status
from pydantic import BaseModel
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy.orm import selectinload

from app.core.permissions import CurrentUser, RequireAdmin
from app.database import get_db
from app.models.laboratorio import Laboratorio, TrabajoLaboratorio

router = APIRouter()


# ─── Schemas ─────────────────────────────────────────────────────────────────

class LaboratorioCreate(BaseModel):
    nombre: str
    telefono: str | None = None
    whatsapp: str | None = None
    email: str | None = None
    contacto: str | None = None
    notas: str | None = None


class LaboratorioUpdate(BaseModel):
    nombre: str | None = None
    telefono: str | None = None
    whatsapp: str | None = None
    email: str | None = None
    contacto: str | None = None
    notas: str | None = None
    activo: bool | None = None


class LaboratorioResponse(BaseModel):
    id: uuid.UUID
    nombre: str
    telefono: str | None
    whatsapp: str | None
    email: str | None
    contacto: str | None
    notas: str | None
    activo: bool
    model_config = {"from_attributes": True}


class TrabajoCreate(BaseModel):
    paciente_id: uuid.UUID
    doctor_id: uuid.UUID
    laboratorio_id: uuid.UUID
    historial_id: uuid.UUID | None = None
    descripcion: str
    pieza_dental: int | None = None
    color: str | None = None
    observaciones: str | None = None
    fecha_salida: date | None = None
    fecha_entrega_prevista: date | None = None
    precio: float | None = None


class TrabajoUpdate(BaseModel):
    laboratorio_id: uuid.UUID | None = None
    descripcion: str | None = None
    pieza_dental: int | None = None
    color: str | None = None
    observaciones: str | None = None
    fecha_salida: date | None = None
    fecha_entrega_prevista: date | None = None
    fecha_recepcion: date | None = None
    fecha_entrega_paciente: date | None = None
    estado: str | None = None
    precio: float | None = None


class PacienteMin(BaseModel):
    id: uuid.UUID
    nombre: str
    apellidos: str
    num_historial: int
    model_config = {"from_attributes": True}


class DoctorMin(BaseModel):
    id: uuid.UUID
    nombre: str
    model_config = {"from_attributes": True}


class TrabajoResponse(BaseModel):
    id: uuid.UUID
    paciente_id: uuid.UUID
    doctor_id: uuid.UUID
    laboratorio_id: uuid.UUID
    historial_id: uuid.UUID | None
    descripcion: str
    pieza_dental: int | None
    color: str | None
    observaciones: str | None
    fecha_salida: date | None
    fecha_entrega_prevista: date | None
    fecha_recepcion: date | None
    fecha_entrega_paciente: date | None
    estado: str
    precio: float | None
    paciente: PacienteMin | None = None
    doctor: DoctorMin | None = None
    laboratorio: LaboratorioResponse | None = None
    model_config = {"from_attributes": True}


ESTADOS_VALIDOS = {
    "pendiente", "enviado", "en_proceso", "recibido", "entregado", "incidencia"
}

# ─── Laboratorios (catálogo) ──────────────────────────────────────────────────

@router.get("/laboratorios", response_model=list[LaboratorioResponse])
async def listar_laboratorios(
    db: Annotated[AsyncSession, Depends(get_db)],
    _: CurrentUser,
    solo_activos: bool = Query(True),
) -> list[LaboratorioResponse]:
    q = select(Laboratorio).order_by(Laboratorio.nombre)
    if solo_activos:
        q = q.where(Laboratorio.activo == True)  # noqa: E712
    result = await db.execute(q)
    return [LaboratorioResponse.model_validate(r) for r in result.scalars().all()]


@router.post("/laboratorios", response_model=LaboratorioResponse, status_code=status.HTTP_201_CREATED)
async def crear_laboratorio(
    data: LaboratorioCreate,
    db: Annotated[AsyncSession, Depends(get_db)],
    _: CurrentUser,
) -> LaboratorioResponse:
    lab = Laboratorio(**data.model_dump())
    db.add(lab)
    await db.commit()
    await db.refresh(lab)
    return LaboratorioResponse.model_validate(lab)


@router.patch("/laboratorios/{lab_id}", response_model=LaboratorioResponse)
async def actualizar_laboratorio(
    lab_id: uuid.UUID,
    data: LaboratorioUpdate,
    db: Annotated[AsyncSession, Depends(get_db)],
    _: CurrentUser,
) -> LaboratorioResponse:
    lab = await db.get(Laboratorio, lab_id)
    if not lab:
        raise HTTPException(status_code=404, detail="Laboratorio no encontrado")
    for field, value in data.model_dump(exclude_none=True).items():
        setattr(lab, field, value)
    await db.commit()
    await db.refresh(lab)
    return LaboratorioResponse.model_validate(lab)


# ─── Trabajos de laboratorio ──────────────────────────────────────────────────

def _trabajo_query():
    return (
        select(TrabajoLaboratorio)
        .options(
            selectinload(TrabajoLaboratorio.paciente),
            selectinload(TrabajoLaboratorio.doctor),
            selectinload(TrabajoLaboratorio.laboratorio),
        )
    )


@router.get("/laboratorio/trabajos", response_model=list[TrabajoResponse])
async def listar_trabajos(
    db: Annotated[AsyncSession, Depends(get_db)],
    _: CurrentUser,
    laboratorio_id: uuid.UUID | None = Query(None),
    paciente_id: uuid.UUID | None = Query(None),
    doctor_id: uuid.UUID | None = Query(None),
    estado: str | None = Query(None),
    pendientes: bool = Query(False),  # solo estados activos (pendiente/enviado/en_proceso)
) -> list[TrabajoResponse]:
    q = _trabajo_query().order_by(TrabajoLaboratorio.created_at.desc())
    if laboratorio_id:
        q = q.where(TrabajoLaboratorio.laboratorio_id == laboratorio_id)
    if paciente_id:
        q = q.where(TrabajoLaboratorio.paciente_id == paciente_id)
    if doctor_id:
        q = q.where(TrabajoLaboratorio.doctor_id == doctor_id)
    if estado:
        q = q.where(TrabajoLaboratorio.estado == estado)
    if pendientes:
        q = q.where(TrabajoLaboratorio.estado.in_(["pendiente", "enviado", "en_proceso"]))
    result = await db.execute(q)
    return [TrabajoResponse.model_validate(t) for t in result.scalars().all()]


@router.post("/laboratorio/trabajos", response_model=TrabajoResponse, status_code=status.HTTP_201_CREATED)
async def crear_trabajo(
    data: TrabajoCreate,
    db: Annotated[AsyncSession, Depends(get_db)],
    _: CurrentUser,
) -> TrabajoResponse:
    trabajo = TrabajoLaboratorio(**data.model_dump())
    db.add(trabajo)
    await db.commit()
    await db.refresh(trabajo)
    result = await db.execute(_trabajo_query().where(TrabajoLaboratorio.id == trabajo.id))
    return TrabajoResponse.model_validate(result.scalar_one())


@router.patch("/laboratorio/trabajos/{trabajo_id}", response_model=TrabajoResponse)
async def actualizar_trabajo(
    trabajo_id: uuid.UUID,
    data: TrabajoUpdate,
    db: Annotated[AsyncSession, Depends(get_db)],
    _: CurrentUser,
) -> TrabajoResponse:
    trabajo = await db.get(TrabajoLaboratorio, trabajo_id)
    if not trabajo:
        raise HTTPException(status_code=404, detail="Trabajo no encontrado")
    cambios = data.model_dump(exclude_none=True)
    if "estado" in cambios and cambios["estado"] not in ESTADOS_VALIDOS:
        raise HTTPException(status_code=422, detail=f"Estado inválido. Válidos: {', '.join(ESTADOS_VALIDOS)}")
    for field, value in cambios.items():
        setattr(trabajo, field, value)
    await db.commit()
    result = await db.execute(_trabajo_query().where(TrabajoLaboratorio.id == trabajo_id))
    return TrabajoResponse.model_validate(result.scalar_one())


@router.delete("/laboratorio/trabajos/{trabajo_id}", status_code=status.HTTP_204_NO_CONTENT, dependencies=[RequireAdmin])
async def eliminar_trabajo(
    trabajo_id: uuid.UUID,
    db: Annotated[AsyncSession, Depends(get_db)],
) -> None:
    trabajo = await db.get(TrabajoLaboratorio, trabajo_id)
    if not trabajo:
        raise HTTPException(status_code=404, detail="Trabajo no encontrado")
    await db.delete(trabajo)
    await db.commit()

File: frontend/src/api/laboratorio.ts
import { apiClient } from "./client";

export interface Laboratorio {
  id: string;
  nombre: string;
  telefono: string | null;
  whatsapp: string | null;
  email: string | null;
  contacto: string | null;
  notas: string | null;
  activo: boolean;
}

export interface LaboratorioCreate {
  nombre: string;
  telefono?: string;
  whatsapp?: string;
  email?: string;
  contacto?: string;
  notas?: string;
}

export type EstadoTrabajo =
  | "pendiente"
  | "enviado"
  | "en_proceso"
  | "recibido"
  | "entregado"
  | "incidencia";

export interface TrabajoLaboratorio {
  id: string;
  paciente_id: string;
  doctor_id: string;
  laboratorio_id: string;
  historial_id: string | null;
  descripcion: string;
  pieza_dental: number | null;
  color: string | null;
  observaciones: string | null;
  fecha_salida: string | null;
  fecha_entrega_prevista: string | null;
  fecha_recepcion: string | null;
  fecha_entrega_paciente: string | null;
  estado: EstadoTrabajo;
  precio: number | null;
  paciente?: { id: string; nombre: string; apellidos: string; num_historial: number } | null;
  doctor?: { id: string; nombre: string } | null;
  laboratorio?: Laboratorio | null;
}

export interface TrabajoCreate {
  paciente_id: string;
  doctor_id: string;
  laboratorio_id: string;
  historial_id?: string;
  descripcion: string;
  pieza_dental?: number;
  color?: string;
  observaciones?: string;
  fecha_salida?: string;
  fecha_entrega_prevista?: string;
  precio?: number;
}

export interface TrabajoUpdate {
  laboratorio_id?: string;
  descripcion?: string;
  pieza_dental?: number;
  color?: string;
  observaciones?: string;
  fecha_salida?: string;
  fecha_entrega_prevista?: string;
  fecha_recepcion?: string;
  fecha_entrega_paciente?: string;
  estado?: EstadoTrabajo;
  precio?: number;
}

export async function getLaboratorios(soloActivos = true): Promise<Laboratorio[]> {
  const r = await apiClient.get("/laboratorios", { params: { solo_activos: soloActivos } });
  return Array.isArray(r.data) ? r.data : [];
}

export async function crearLaboratorio(data: LaboratorioCreate): Promise<Laboratorio> {
  const r = await apiClient.post("/laboratorios", data);
  return r.data;
}

export async function actualizarLaboratorio(id: string, data: Partial<LaboratorioCreate & { activo: boolean }>): Promise<Laboratorio> {
  const r = await apiClient.patch(`/laboratorios/${id}`, data);
  return r.data;
}

export async function getTrabajos(params?: {
  laboratorio_id?: string;
  paciente_id?: string;
  doctor_id?: string;
  estado?: EstadoTrabajo;
  pendientes?: boolean;
}): Promise<TrabajoLaboratorio[]> {
  const r = await apiClient.get("/laboratorio/trabajos", { params });
  return Array.isArray(r.data) ? r.data : [];
}

export async function crearTrabajo(data: TrabajoCreate): Promise<TrabajoLaboratorio> {
  const r = await apiClient.post("/laboratorio/trabajos", data);
  return r.data;
}

export async function actualizarTrabajo(id: string, data: TrabajoUpdate): Promise<TrabajoLaboratorio> {
  const r = await apiClient.patch(`/laboratorio/trabajos/${id}`, data);
  return r.data;
}

export async function eliminarTrabajo(id: string): Promise<void> {
  await apiClient.delete(`/laboratorio/trabajos/${id}`);
}

/** Abre WhatsApp con el número del laboratorio y un mensaje prefabricado. */
export function abrirWhatsappLab(whatsapp: string, mensaje: string): void {
  const numero = whatsapp.replace(/\D/g, "");
  const url = `https://wa.me/${numero}?text=${encodeURIComponent(mensaje)}`;
  window.open(url, "_blank");
}
